Include data for API calls in URL to maximize compatibility

profiluefter opened the pull request:
According to the [official moodle api documentation][1] one should include all the necessary parameters for the API call in the URL.
This commit makes it possible to use moodle-dl with the austrian [eduvidual][2] moodle service.

[1]: https://docs.moodle.org/dev/Creating_a_web_service_client#Simple_REST_request_example
[2]: https://www.eduvidual.at/my/

C0D3D3V commented:
One thing that I do not like, is that the token will now be included in the responses log if you use `moodle-dl --log-responses`. Because sometimes I request this log from people that have problems or request a new feature... and I do not want to know their tokens.

Would be nice if you add an filter for that, or maybe you have an better idea.

profiluefter commented:
I think a filter is a good idea, I'll look into it

profiluefter commented:
I added a regex which matches everything that looks like a token instead of searching for the token specifically so that even if the token changes at runtime (I'm not sure if thats currently the case or maybe in the future) it should work.

C0D3D3V commented:
I found the change, it was when we switched from Get to Post.
 https://github.com/C0D3D3V/Moodle-Downloader-2/commit/c96dc38c7c7b9bbafcd92707611d37251370b375

Someone complained that his sensible data could be logged in the http server log, because this log contains the called URLs. Thats why we switched to Post, and then also removed the token from the URL because it worked nonetheless.

C0D3D3V commented:
For me the downloader works on your moodle, without adding the token to the URL. I could sucessfully download this course: https://www.eduvidual.at/course/view.php?id=21408
see: 
![grafik](https://user-images.githubusercontent.com/14315968/176138967-4f29d71f-3787-4b5b-872f-5aa84a6eb933.png)


But maybe they messed up for sso logins, I only used the normal login for externals.

C0D3D3V commented:
I now also tested it with SSO over google. And it worked the same way. Maybe we could have a call in discord where you show me that it does not work without this PR.

C0D3D3V commented on moodle_dl/moodle_connector/request_helper.py:
```diff
@@ -173,18 +173,31 @@ def post_REST(self, function: str, data: {str: str} = None, timeout: str = 60) -
+        data = {'moodlewsrestformat': 'json'}
+
+        if data_obj is not None:
+            data.update(data_obj)
```
Why did you also add all other data to the URL? 
I really do not think this is worth it. I will look up the moodle code. I guess adding the token to the URL is not worth it too.

C0D3D3V commented:
This function: 

https://github.com/moodle/moodle/blob/5500d143f406baa5046fd03a00e88e4cfc5f4bb9/webservice/rest/locallib.php#L73-L104

 is used to get the token for webservice calls to the rest endpoint.

And also this: https://github.com/moodle/moodle/blob/5500d143f406baa5046fd03a00e88e4cfc5f4bb9/lib/moodlelib.php#L666-L687

for get_site_info
https://github.com/moodle/moodle/blob/5500d143f406baa5046fd03a00e88e4cfc5f4bb9/webservice/externallib.php#L106

And both get the token from Post Body and URL Get Parameter. 

So there should really be no need to set it in the URL. 
I guess it does not harm anyone, but it is also useless. 

It is only required if we would switch to the soap or xmlrpc endpoint.

profiluefter commented:
I think it's totally possible that it is just an issue specific to eduvidual as it is a very heavily modified version of moodle.

> For me the downloader works on your moodle, without adding the token to the URL. I could sucessfully download this course: https://www.eduvidual.at/course/view.php?id=21408 see: ![grafik](https://user-images.githubusercontent.com/14315968/176138967-4f29d71f-3787-4b5b-872f-5aa84a6eb933.png)
> 
> But maybe they messed up for sso logins, I only used the normal login for externals.

That course is also available without login so maybe that's why it works for you. However I did also have issues with passing parameters to the methods.

Default request with parameters and token in POST body (doesn't work):
![grafik](https://user-images.githubusercontent.com/18028031/176266604-a50c34e2-4e3b-4fd5-bb6a-005d8999f652.png)

With token in body and query but parameters only in body (missing parameters!):
![grafik](https://user-images.githubusercontent.com/18028031/176266805-59c897d1-c684-43b2-9efb-396d157064d3.png)

With both parameters and token in both body and query (does work):
![grafik](https://user-images.githubusercontent.com/18028031/176267162-c33626a0-5121-43a0-b312-c1cfbe5841db.png)

Above also works if I remove the POST body. Eduvidual seemingly doesn't care about POST parameters.

I, of course, believe you that moodle-dl has worked with the previous API configuration with only post parameters however according to the [simple API example from moodle docs](https://docs.moodle.org/dev/Creating_a_web_service_client#Simple_REST_request_example) one should use query parameters and Eduvidual enforces that I guess.
Maybe some sort of compatibility mode would be better but I think putting the parameters in both is still the better idea. 

> Someone complained that his sensible data could be logged in the http server log, because this log contains the called URLs. Thats why we switched to Post, and then also removed the token from the URL because it worked nonetheless.

I'm assuming you're talking about the server side log file which shouldn't be an attack vector because the server already knows the sensitive data? Or if we're talking about reverse proxy situations then the server could also inspect the request/response body. Sensitive data logging should be prevented by configuring the server appropriately so imho this shouldn't be an issue of this software. Maybe I could add a "paranoid mode" :) or something like that instead of an eduvidual user having to customize the config?

Also I think you wrote on discord that the moodle app also puts it in both then I guess that's already the accepted level of privacy/security? Also one might argue that it's also an improvement to better imitate the official app.

> Why did you also add all other data to the URL?
> I really do not think this is worth it. I will look up the moodle code. I guess adding the token to the URL is not worth it too.

Sorry btw for not explaining it before but I just added it so it works with my moodle.

So what are your thoughts on this or do you have an idea why it doesn't work with Eduvidual?
Btw I tested the above errors with many different api methods.
